### PR TITLE
feat: Added add card component in dashboard

### PR DESCRIPTION
**Pull request by OmkarJ13**
[workerB](https://workerb.linearb.io/v2/badge/collaboration-page?magicLinkId=xC0_Dbv)
### <samp>🤖 Generated by Copilot at a2a743a</samp>

Added a new feature to show card stats and allow adding cards from the dashboard. Refactored and cleaned up the code and styles for the `stats` and `spentCards` components. Added a new component `app-add-card` and its tests. Removed unused code from `src/global.scss`.

### <samp>🤖 Generated by Copilot at a2a743a</samp>

> _`app-add-card` shows_
> _button to swipe and add cards_
> _empty state in spring_

### <samp>🤖 Generated by Copilot at a2a743a</samp>

*  Refactor the code to use a new component `app-add-card` to show the add card empty state in the card stats section ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-06339ae5900c441920a2a6c20ee72bb01b2130e95691951c2ed1d3ef10bdcf5eL154-R265), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-78492bd8c7b8576f817b06758cd4ff3d0b54df406016f5c86896d184221ebeaeR15-R19), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-5ffddeffb7e5e557ebf833371417195723bbe379cb2654166d6183c3cd6fa22cR7-R14), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-1a597830af6ef3cf160383a46c39c9204cae7151ff583644a3c58e8946564ab1L54-R70), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-71ee557103ff1e53a7c4f46a8d510ceb4134b7e22caa6c531c2255b76317100aR141), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-71ee557103ff1e53a7c4f46a8d510ceb4134b7e22caa6c531c2255b76317100aR259), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-71ee557103ff1e53a7c4f46a8d510ceb4134b7e22caa6c531c2255b76317100aR382), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-de6c7d3b1be29b1fbdc515c9449a7d3779efcf15c279a1e8b88a2585ba47f818L1101-L1106))
*  Remove some unused classes and ngIf conditions from the HTML template of the `StatsComponent` ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-06339ae5900c441920a2a6c20ee72bb01b2130e95691951c2ed1d3ef10bdcf5eL154-R265))
*  Remove some unused imports and add some new imports for the card stats feature in the `StatsComponent` ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L1-R12))
*  Add some new properties to the `StatsComponent` to store the flags for whether the CCC, yodlee, and personal cards features are enabled ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72R19-R23))
*  Inject a new service `orgUserSettingsService` to the `StatsComponent` to fetch the user-specific settings for the card features ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L75-R93))
*  Add type annotations to the class properties and methods of the `StatsComponent` for better readability and type checking ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72R71-R76), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L90-R101), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L125-R136), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L141-R153), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L150-R172), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L167-R183), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L177-R210), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L199-R227), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L215-R248), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L232-R260), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L250-R278))
*  Fix a bug where the `isCCCStatsLoading` flag was not set to false after the observable returned by `getCCCDetails` completed ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L150-R172))
*  Use the `cloneDeep` function from lodash to create a deep copy of the card names array instead of using JSON.parse and JSON.stringify ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L141-R153))
*  Use the `forkJoin` operator to combine the observables returned by `orgSettingsService.get` and `orgUserSettingsService.get`, and set the flags for the card features based on the settings ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L177-R210))
*  Call the `initializeCCCStats` method only if the CCC feature is enabled ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-2821ecdc9d0f1aa4b5588d4dca7a7c2335690a4a378372ac782c9cfe41d78f72L177-R210))
*  Fix a minor padding issue in the `stats--expenses-stats-row` class ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-0e2e75bc5de76669e02166b7b5e28d2626e8daa795b34a8a66422408e2fb0eb7L152-R144))
*  Rename the class `stats--ccc-cards-header` to `stats--ccc-stats-header` for consistency, and add a new class `stats--ccc-zero-state` for styling the add card empty state ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-0e2e75bc5de76669e02166b7b5e28d2626e8daa795b34a8a66422408e2fb0eb7L162-R162))
*  Create and style the new component `app-add-card` that displays a button to add a new card with an optional zero state message, and emits an output event when the button is clicked ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-f9e2659014c7cb2b02a16174e456a776b4765d2278ce91e38b858dc7bc39eacdR1-R9), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-ecbafe954d64617bf544646bc018f65145c11265cf634cc7b2eb58ff299e3989R1-R28), [link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-8786d99e467a42f3414231036f3565bda415e6fc2fa3a2eb5c95c30586c41d9fR1-R12))
*  Write unit tests for the new component `app-add-card` using the TestBed and ComponentFixture utilities from Angular ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-ae3550de5730e363e5a41cba3d785756a29348edf5ffa1e339de66b90b52da67R1-R53))
*  Update the unit test for the `spentCards` property of the `SpentCardsComponent` to account for the new swiper slide that contains the `app-add-card` component ([link](https://github.com/fylein/fyle-mobile-app/pull/2220/files?diff=unified&w=0#diff-1a597830af6ef3cf160383a46c39c9204cae7151ff583644a3c58e8946564ab1L35-R36))

## Clickup
https://app.clickup.com/t/85ztm6eje
https://app.clickup.com/t/85ztm6e2t

## Code Coverage
Please add code coverage here

## UI Preview
image
image

## Figma Designs
https://www.figma.com/file/qDAvQycE6BhZ6ABmXLeVWL/%5B2023Q2%5D-Unified-Card-Enrollment-for-Mobile?node-id=2%3A6&mode=dev

### CODE REVIEW

**Review by spudmoore — APPROVED** [review]

File: src/app/fyle/dashboard/stats/stats.component.ts [hunk 1]
```diff
@@ -1,22 +1,26 @@
-import { Component, EventEmitter, OnInit, Output } from '@angular/core';
```
Comment (OmkarJ13) [inline src/app/fyle/dashboard/stats/stats.component.ts:1]: Bunch of lint fixes in this file

File: src/global.scss [hunk 2]
```diff
@@ -1098,12 +1098,6 @@ ion-popover.delete-button-class {
   opacity: 0.3;
 }
 
-.spent-cards {
```
Comment (OmkarJ13) [inline src/global.scss:1101]: This was adding a negative margin to each active swiper, it was making the previous slide look odd, so have fixed this by removing it.

File: src/app/shared/components/spent-cards/spent-cards.component.scss [hunk 3]
```diff
@@ -4,3 +4,11 @@
   padding: 12px 0;
   height: 140px;
 }
+
+.spent-cards--add-card {
```
Comment (OmkarJ13) [inline src/app/shared/components/spent-cards/spent-cards.component.scss:8]: Followed a similar positioning and dimensions as we use for other slides in the swiper

File: src/app/fyle/dashboard/stats/stats.component.ts [hunk 4]
```diff
@@ -138,48 +149,65 @@ export class StatsComponent implements OnInit {
...
 
-  initializeCCCStats() {
-    this.dashboardService.getCCCDetails().subscribe((details) => {
-      this.cardTransactionsAndDetails = this.getCardDetail(details.cardDetails);
-    });
-    finalize(() => (this.isCCCStatsLoading = false));
+  initializeCCCStats(): void {
+    this.dashboardService
+      .getCCCDetails()
+      .pipe(finalize(() => (this.isCCCStatsLoading = false)))
```
Comment (OmkarJ13) [inline src/app/fyle/dashboard/stats/stats.component.ts:169]: Fixed the finalize call here, it should be called in pipe

File: src/app/fyle/dashboard/stats/stats.component.scss [hunk 5]
```diff
@@ -3,14 +3,6 @@
...
 
 .stats {
-  &--ccc-cards-grid {
-    margin-bottom: 2%;
-  }
-
-  &--ccc-cards-grid-multiple {
-    margin-bottom: -4%;
-  }
-
```
Comment (OmkarJ13) [inline src/app/fyle/dashboard/stats/stats.component.scss:13]: Not needed anymore

File: src/app/fyle/dashboard/stats/stats.component.ts [hunk 6]
```diff
@@ -138,48 +149,65 @@ export class StatsComponent implements OnInit {
...
+
+      this.isPersonalCardsEnabled =
+        orgSettings.org_personal_cards_settings.allowed &&
+        orgSettings.org_personal_cards_settings.enabled &&
+        orgUserSettings.personal_cards_settings.enabled;
+
+      if (this.isCCCEnabled) {
         that.isCCCStatsLoading = true;
         that.initializeCCCStats();
-      } else {
```
Comment (OmkarJ13) [inline src/app/fyle/dashboard/stats/stats.component.ts:181]: This is not needed

File: src/global.scss (hunk 2 above)
Comment (OmkarJ13) [reply src/global.scss:1101]: Before: 
image

After:

File: src/app/fyle/dashboard/stats/stats.component.scss [hunk 7]
```diff
@@ -149,7 +141,7 @@ $ccc-bank-name: #c4cae8;
 
   &--expenses-stats-row {
     padding-bottom: 18px;
-    padding-top: 19px;
```
Comment (OmkarJ13) [inline src/app/fyle/dashboard/stats/stats.component.scss:152]: Corrected odd value, it should be 18

File: src/app/fyle/dashboard/stats/stats.component.html [hunk 8]
```diff
@@ -151,120 +151,118 @@
...
+      <ng-container *ngIf="cardTransactionsAndDetails?.length === 0 && (isYodleeEnabled || isPersonalCardsEnabled)">
+        <div class="stats--header stats--ccc-stats-header">CARD SPEND</div>
+        <div class="stats--ccc-zero-state">
+          <app-add-card [showZeroStateMessage]="true"></app-add-card>
+        </div>
+      </ng-container>
     </div>
+
+    <ng-template #cccStatsLoading>
+      <div class="stats--ccc-card-details">
```
Comment (Julias0) [inline src/app/fyle/dashboard/stats/stats.component.html:186]: should we move the skeleton loader into a different component? This is taking up a bulk of the template space

**Review by Julias0 — APPROVED** [review]

File: src/app/fyle/dashboard/stats/stats.component.html (hunk 8 above)
Comment (OmkarJ13) [reply src/app/fyle/dashboard/stats/stats.component.html:186]: Yes, thinking of moving the skeleton loader + the cards section into a different component, will take care of this in a separate PR

**Pull request merged**